File: laravelProject/app/Application/Services/VehiculoService.php
<?php

namespace App\Application\Services;

use App\Domain\Entities\Vehiculo;
use App\Domain\Repositories\MarcaRepositoryInterface;
use App\Domain\Repositories\PersonaRepositoryInterface;
use App\Domain\Repositories\VehiculoRepositoryInterface;
use Illuminate\Database\Eloquent\ModelNotFoundException;

/**
 * Servicio de aplicación para Vehículo.
 *
 * Contiene la lógica de negocio para operaciones con vehículos.
 * Sigue el principio de responsabilidad única (SOLID).
 */
class VehiculoService
{
    public function __construct(
        private VehiculoRepositoryInterface $vehiculoRepository,
        private MarcaRepositoryInterface $marcaRepository,
        private PersonaRepositoryInterface $personaRepository
    ) {
    }

    /**
     * Crea un nuevo vehículo.
     *
     * @param array<string, mixed> $data
     * @return Vehiculo
     * @throws ModelNotFoundException Si la marca no existe
     */
    public function create(array $data): Vehiculo
    {
        // Validar que la marca exista
        $marca = $this->marcaRepository->findById($data['marca_id']);
        if (!$marca) {
            throw new ModelNotFoundException('Marca no encontrada.');
        }
        
        $vehiculo = Vehiculo::fromArray($data);
        
        return $this->vehiculoRepository->save($vehiculo);
    }

    /**
     * Obtiene todos los vehículos.
     *
     * @param int $skip
     * @param int $limit
     * @return array<Vehiculo>
     */
    public function list(int $skip = 0, int $limit = 100): array
    {
        return $this->vehiculoRepository->findAll($skip, $limit);
    }

    /**
     * Obtiene un vehículo por ID.
     *
     * @param int $id
     * @return Vehiculo
     * @throws ModelNotFoundException
     */
    public function get(int $id): Vehiculo
    {
        $vehiculo = $this->vehiculoRepository->findById($id);
        
        if (!$vehiculo) {
            throw new ModelNotFoundException('Vehículo no encontrado.');
        }
        
        return $vehiculo;
    }

    /**
     * Actualiza un vehículo.
     *
     * @param int $id
     * @param array<string, mixed> $data
     * @return Vehiculo
     * @throws ModelNotFoundException
     */
    public function update(int $id, array $data): Vehiculo
    {
        $vehiculo = $this->get($id);
        
        // Si se está cambiando la marca, validar que exista
        if (isset($data['marca_id']) && $data['marca_id'] !== $vehiculo->marcaId) {
            $marca = $this->marcaRepository->findById($data['marca_id']);
            if (!$marca) {
                throw new ModelNotFoundException('Marca no encontrada.');
            }
        }
        
        // Actualizar campos
        foreach ($data as $key => $value) {
            if ($key === 'modelo') {
                $vehiculo->modelo = $value;
            } elseif ($key === 'marca_id') {
                $vehiculo->marcaId = $value;
            } elseif ($key === 'numero_puertas') {
                $vehiculo->numeroPuertas = $value;
            } elseif ($key === 'color') {
                $vehiculo->color = $value;
            }
        }
        
        return $this->vehiculoRepository->save($vehiculo);
    }

    /**
     * Elimina un vehículo.
     *
     * @param int $id
     * @return bool
     * @throws ModelNotFoundException
     */
    public function delete(int $id): bool
    {
        $this->get($id); // Valida que exista
        
        return $this->vehiculoRepository->delete($id);
    }

    /**
     * Obtiene los vehículos de una persona.
     *
     * @param int $personaId
     * @return array<Vehiculo>
     * @throws ModelNotFoundException
     */
    public function getVehiculosByPersona(int $personaId): array
    {
        // Validar que la persona exista
        $persona = $this->personaRepository->findById($personaId);
        if (!$persona) {
            throw new ModelNotFoundException('Persona no encontrada.');
        }
        
        return $this->vehiculoRepository->findByPersonaId($personaId);
    }

    /**
     * Asigna un propietario a un vehículo.
     *
     * @param int $vehiculoId
     * @param int $personaId
     * @return Vehiculo
     * @throws ModelNotFoundException
     * @throws \Exception Si el propietario ya está asignado
     */
    public function addPropietario(int $vehiculoId, int $personaId): Vehiculo
    {
        // Validar que el vehículo exista
        $vehiculo = $this->get($vehiculoId);
        
        // Validar que la persona exista
        $persona = $this->personaRepository->findById($personaId);
        if (!$persona) {
            throw new ModelNotFoundException('Persona no encontrada.');
        }
        
        // Intentar agregar el propietario
        $added = $this->vehiculoRepository->addPropietario($vehiculoId, $personaId);
        if (!$added) {
            throw new \Exception('El propietario ya está asociado a este vehículo.');
        }
        
        return $this->get($vehiculoId);
    }
}

File: fastApiProject/app/application/services/vehiculo_service.py
from __future__ import annotations

from typing import List

from fastapi import HTTPException, status

from app.domain.entities import Vehiculo
from app.domain.repositories import MarcaRepository, PersonaRepository, VehiculoRepository


class VehiculoService:
    def __init__(
        self,
        vehiculo_repository: VehiculoRepository,
        marca_repository: MarcaRepository,
        persona_repository: PersonaRepository,
    ) -> None:
        self._vehiculo_repository = vehiculo_repository
        self._marca_repository = marca_repository
        self._persona_repository = persona_repository

    def create(self, vehiculo: Vehiculo) -> Vehiculo:
        self._ensure_marca_exists(vehiculo.marca_id)
        return self._vehiculo_repository.create(vehiculo)

    def list(self, skip: int, limit: int) -> List[Vehiculo]:
        return self._vehiculo_repository.list(skip, limit)

    def get(self, vehiculo_id: int) -> Vehiculo:
        vehiculo = self._vehiculo_repository.get(vehiculo_id)
        if not vehiculo:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Vehículo no encontrado.",
            )
        return vehiculo

    def update(self, vehiculo_id: int, data: dict) -> Vehiculo:
        if "marca_id" in data:
            self._ensure_marca_exists(data["marca_id"])
        if "propietarios_ids" in data:
            for persona_id in data["propietarios_ids"]:
                self._ensure_persona_exists(persona_id)
            return self._vehiculo_repository.set_propietarios(
                vehiculo_id, data.pop("propietarios_ids")
            )
        return self._vehiculo_repository.update(vehiculo_id, data)

    def delete(self, vehiculo_id: int) -> None:
        self.get(vehiculo_id)
        self._vehiculo_repository.delete(vehiculo_id)

    def add_propietario(self, vehiculo_id: int, persona_id: int) -> Vehiculo:
        self.get(vehiculo_id)
        self._ensure_persona_exists(persona_id)
        try:
            return self._vehiculo_repository.add_propietario(vehiculo_id, persona_id)
        except ValueError as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=str(exc),
            ) from exc

    def _ensure_marca_exists(self, marca_id: int) -> None:
        marca = self._marca_repository.get(marca_id)
        if not marca:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, detail="Marca no encontrada."
            )

    def _ensure_persona_exists(self, persona_id: int) -> None:
        persona = self._persona_repository.get(persona_id)
        if not persona:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, detail="Persona no encontrada."
            )

